internal/controllers/machine/machine_controller_kubelet_serving_csr.go

```diff
@@ -0,0 +1,173 @@
+	NodesPrefix = "system:node:"
+
+	// NodesGroup defines the group name for a node.
+	NodesGroup = "system:nodes"
+)
+
+func (r *Reconciler) reconcileKubletServingCertificateSigningRequest(ctx context.Context, cluster *clusterv1.Cluster, machine *clusterv1.Machine) (ctrl.Result, error) {
+	log := ctrl.LoggerFrom(ctx, "machine", machine.Name, "namespace", machine.Namespace)
+	log = log.WithValues("cluster", cluster.Name)
+
+	// TODO can NodeRef be nil at this point?
+	nodeRef := machine.Status.NodeRef
+
+	remoteClient, err := r.Tracker.GetClient(ctx, util.ObjectKey(cluster))
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	restConfig, err := remote.RESTConfig(ctx, controllerName, r.Client, util.ObjectKey(cluster))
+	if err != nil {
+		return ctrl.Result{}, nil
+	}
+	kubeClient, err := kubernetes.NewForConfig(restConfig)
+
+	node := corev1.Node{}
+	if err := remoteClient.Get(ctx, client.ObjectKey{Name: nodeRef.Name}, &node); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Find the corresponding CSRs for this node
+	// TODO: Add filter options
+	csrList := certificatesv1.CertificateSigningRequestList{}
+	if err := remoteClient.List(ctx, &csrList, &client.ListOptions{
+		FieldSelector: fields.AndSelectors(
+			fields.OneTermEqualSelector("spec.username", NodesPrefix+node.Name),
+			fields.OneTermEqualSelector("spec.signer", KubeletServingSignerName),
+		),
+	}); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	for _, csr := range csrList.Items {
+
+		block, _ := pem.Decode(csr.Spec.Request)
+		if block.Type != "CERTIFICATE REQUEST" {
+			return ctrl.Result{}, fmt.Errorf("Unexpected PEM type")
+		}
+		cr, err := x509.ParseCertificateRequest(block.Bytes)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+
+		// TODO: Does this case happen where we have a CSR whose contents dont match the object?
+		if cr.Subject.CommonName != csr.Spec.Username {
+			return ctrl.Result{}, fmt.Errorf("csr Username didnt match CN")
+		}
+		for i, org := range cr.Subject.OrganizationalUnit {
+			if org != csr.Spec.Groups[i] {
+				return ctrl.Result{}, fmt.Errorf("csr Groups didn't match O")
+			}
+		}
+
+		condition := certificatesv1.CertificateSigningRequestCondition{
+			LastUpdateTime: metav1.Time{Time: time.Now()},
+		}
+		if err := validateCSR(cr, machine.Status.Addresses); err != nil {
+			// TODO: we do not want to fail early, just skip this CSR
+			condition.Type = certificatesv1.CertificateDenied
+			condition.Reason = "CSRValidationFailed"
+			condition.Status = "True"
+			condition.Message = fmt.Sprintf("Validation failed: %s", err)
+		} else {
+			condition.Type = certificatesv1.CertificateApproved
+			condition.Reason = "CSRValidationSucceed"
+			condition.Status = "True"
+			condition.Message = "Validation was successful"
+
+		}
+		csr.Status.Conditions = append(csr.Status.Conditions, condition)
+
+		if _, err := kubeClient.CertificatesV1().CertificateSigningRequests().UpdateApproval(ctx, csr.Name, &csr, metav1.UpdateOptions{}); err != nil {
```

Comment:
This should be somehow composable. We should not break existing workflows and preclude existing ad-hoc approvers from being more restrictive, by approving explicitly here we are effectively defeating any existing approval workflow.